rabbitmq/src/test/java/io/micronaut/rabbitmq/intercept/EnvironmentLoader.java

```diff
@@ -0,0 +1,130 @@
+package io.micronaut.rabbitmq.intercept;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import com.github.dockerjava.api.model.HealthCheck;
+import io.micronaut.context.env.ActiveEnvironment;
+import io.micronaut.context.env.PropertySource;
+import io.micronaut.context.env.PropertySourceLoader;
+import io.micronaut.core.io.ResourceLoader;
+import org.awaitility.Awaitility;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.BindMode;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.InternetProtocol;
+import org.testcontainers.containers.Network;
+import org.testcontainers.containers.output.Slf4jLogConsumer;
+import org.testcontainers.containers.wait.strategy.AbstractWaitStrategy;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.utility.DockerImageName;
+
+/**
+ * Abuse the fact that properties are loaded before the application context is initialized to setup the
+ * RabbitMQ test cluster.
+ */
+public class EnvironmentLoader implements PropertySourceLoader {
+    private static final int AMQP_PORT = 5672;
+    private static final int MANAGEMENTUI_PORT = 15672;
+    public static final String EXCHANGE = "test-exchange";
+    public static final String QUEUE = "test-durable-queue";
+    public static final String NODE1 = "node1";
+    public static final String NODE2 = "node2";
+    public static final String NODE3 = "node3";
+    public static final int NODE1_PORT = AMQP_PORT;
+    public static final int NODE2_PORT = AMQP_PORT + 1;
+    public static final int NODE3_PORT = AMQP_PORT + 2;
+    public static final String NODE1_CONT = "containerNode1";
+    public static final String NODE2_CONT = "containerNode2";
+    public static final String NODE3_CONT = "containerNode3";
+
+    private static final Logger log = LoggerFactory.getLogger(EnvironmentLoader.class);
+    private static final DockerImageName RABBIT_IMAGE = DockerImageName.parse("rabbitmq:3-management");
+    private static final String CLUSTER_COOKIE = "test-cluster";
+    private static final String RABBIT_CONFIG_PATH = ClassLoader.getSystemResource("rabbit/rabbitmq.conf").getPath();
+    private static final String RABBIT_DEFINITIONS_PATH = ClassLoader.getSystemResource("rabbit/definitions.json")
+            .getPath();
+    private static final Network mqClusterNet = Network.newNetwork();
+
+    private static Map<String, Object> cache = new HashMap<>();
+
+    @Override
+    public Optional<PropertySource> load(String resourceName, ResourceLoader resourceLoader) {
+        if (cache.isEmpty()) {
+            cache.put(NODE1_CONT, new GenericContainer<>(RABBIT_IMAGE));
```

Comment:
Creating containers in a property source loader is a hack. Why cant this be done in an abstract class like the other tests?

Reply:
This hack causes the test extension to wait until the cluster is running. Micronaut test tries to start the application context right on the start of the test. It does not wait for the RabbitMQ cluster containers to complete their startup. So the connection attempts fail.

I tried to put the cluster start in a static initializer in an abstract class, so it gets definitely executed before the test class is initialized. But still, the Micronaut test extension starts the application context in parallel and everything fails.

How can i ensure that the cluster is running before the application context becomes initialized?

Reply:
If you tell me that this issue gets solved by using Spock... then i will definitely have a look into it. :)

Reply:
> Micronaut test tries to start the application context right on the start of the test

You don't have to use Micronaut test. I will convert the test to use spock and create the containers the current way. Do I have privileges to push to your branch?

Reply:
I hope you did not invest any work yet. I took the chance to get familiar with Spock and converted the test.
In the end it was easier with Spock than with JUnit.

Reply:
@mscheibler No I haven't. Glad you were able to do it! I hope you enjoy Spock and use it elsewhere. I'll give this another review soon